weeks/week02/section04.md

```diff
@@ -1,32 +1,289 @@
+A `set` is an `unordered` and `unindexed` collection of `unique` elements. They are useful when you need to eliminate `duplicates`, perform set operations like `unions`, `intersections`, and `differences`, and ensure that each element appears only once. Sets are `mutable`, but the elements themselves must be `immutable`.
 
 
 ## References:
 
-1. [Sets - Python Documentation](https://docs.python.org/3/tutorial/datastructures.html#sets)
-2. [Python Sets - Real Python](https://realpython.com/python-sets/)
+- [Python Official Documentation on Sets](https://docs.python.org/3/library/stdtypes.html#set-types-set-frozenset)
+- [Real Python - Python Sets: A Detailed Guide](https://realpython.com/python-sets/)
+- [W3Schools - Python Set](https://www.w3schools.com/python/python_sets.asp)
 
+---
 
-## 4.1 Understanding Sets
 
-Sets are used to store multiple items in a single variable, but unlike lists, they do not allow duplicates.
+## 4.1 Creating
+
+To create a `set` in Python, you can either define it with curly braces `{}` containing elements or use the `set()` function for an empty `set` or when initializing from an iterable.
 
 
 ### Syntax:
 
 ```python
-my_set = {1, 2, 3}
+my_set = {immutable_element_1, immutable_element_2, immutable_element_3}
 ```
 
 
 ### Example:
 
 ```python
-unique_numbers = {1, 2, 3, 2}
-print(unique_numbers)  # Outputs: {1, 2, 3}
+# Creating a set
+test_results = {'pass', 'fail', 'error', 'pass'} # Output: {'pass', 'error', 'fail'}
+
+# Creating an empty set
+empty_set = set() # Output: set()
+```
+
+
+## 4.2 Reading
+
+Reading elements in a `set` is typically done through iteration. Since sets are `unordered`,  you cannot access elements by `index`.
+
+
+### Using `in`:
+
+`in` checks for membership.
+
+```python
+colors = {'red', 'blue', 'green'}
+
+# Reading elements in a set
+if 'blue' in colors:
+    print('Blue is present in the set')
+
+# Iterating over a set
+for color in colors:
+    print(color)
+
+print(colors[0])  # Raises TypeError: 'set' object is not subscriptable
+```
+
+
+### Using `len()`
+
+`len()` returns the number of elements in a set.
+
+```python
+my_set = {1, 2, 3}
+len(my_set)  # 3
+```
+
+
+## 4.3 Updating
+
+You can use the `add()` and `update()` methods to add elements to a `set`.
+
+
+### `add()`:
+
+It only adds a single element to a `set`. Adding existing element will be ignored
+
+```python
+colors = {'red', 'blue', 'green'}
+colors.add('yellow')  # `colors` now is {'red', 'blue', 'yellow', 'green'}
+
+# Adding an existing element will be ignored
+colors.add('yellow')  # `colors` now is {'red', 'blue', 'yellow', 'green'}
+```
+
+
+### `update()`
+
+It adds multiple elements. Adding existing elements will be ignored
+
+```python
+colors = {'red', 'blue', 'green', 'yellow'}
+colors.update(['pink', 'black', 'red', 'yellow']) # `colors` now is {'yellow', 'pink', 'green', 'black', 'red', 'blue'}
+```
+
+
+### Shallow copy
```

Comment:
Explanation what is 'Shallow' copy and what is deep copy is needed (or link to the corresponding web page). Even if it is mentioned earlier. It should be independent page.

Reply:
@SSN-SPB @ericrommel I added a ref in dict PR
![image](https://github.com/user-attachments/assets/d2dc6b09-2444-4de1-b092-526945cbc04f)